File: skills/novel-writer/scripts/validate_patch.py
#!/usr/bin/env python3
import argparse
import json
from pathlib import Path
from typing import Any, Dict, List

from _novel_lib import load_json, memory_path


def _by_name(items: List[Dict[str, Any]]) -> Dict[str, Dict[str, Any]]:
    return {str(item.get("name", "")).strip(): item for item in items if item.get("name")}


def _validate_characters(current: Dict[str, Any], patch: Dict[str, Any]) -> None:
    items = current.get("characters", [])
    existing = _by_name(items)

    add = patch.get("add", []) or []
    update = patch.get("update", []) or []
    delete = patch.get("delete", []) or []

    to_delete = {str(item.get("name", item)).strip() for item in delete if item}
    protected = {n for n, v in existing.items() if v.get("protected") is True}
    illegal = protected.intersection(to_delete)
    if illegal:
        raise ValueError(f"cannot delete protected characters: {sorted(illegal)}")

    # Guardrail: major character shrink protection
    majors_before = [v for v in existing.values() if v.get("role") == "main" or v.get("protected")]
    majors_after = list(existing.values())

    # Apply adds/updates to compute majors_after
    for item in add:
        name = str(item.get("name", "")).strip()
        if not name:
            continue
        existing[name] = {**existing.get(name, {}), **item}
    for item in update:
        name = str(item.get("name", "")).strip()
        if not name:
            continue
        existing[name] = {**existing.get(name, {}), **item}
    for name in list(to_delete):
        if name in existing and name not in protected:
            existing.pop(name, None)

    majors_after = [v for v in existing.values() if v.get("role") == "main" or v.get("protected")]
    if majors_before and len(majors_after) < max(1, int(len(majors_before) * 0.7)):
        raise ValueError("character shrink validation failed (major characters drop)")


def _validate_world(current: Dict[str, Any], patch: Dict[str, Any]) -> None:
    protected = set(current.get("protected_keys", []))
    delete = patch.get("delete", []) or []
    illegal = [key for key in delete if key in protected]
    if illegal:
        raise ValueError(f"cannot delete protected world keys: {sorted(illegal)}")


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--project", required=True)
    ap.add_argument("--patch", required=True)
    args = ap.parse_args()

    patch_path = Path(args.patch)
    if not patch_path.exists():
        raise SystemExit(f"patch file not found: {patch_path}")

    patch = json.loads(patch_path.read_text())

    characters = load_json(memory_path(args.project, "characters"), {"characters": []})
    world = load_json(memory_path(args.project, "world_bible"), {"world": {}, "protected_keys": []})

    _validate_characters(characters, patch.get("characters", {}))
    _validate_world(world, patch.get("world_bible", {}))

    print(json.dumps({"status": "ok"}))


if __name__ == "__main__":
    main()

File: skills/novel-writer/ts/validate-patch.ts
#!/usr/bin/env npx tsx
/**
 * validate-patch.ts
 * Validate a patch JSON before applying it to memory.
 * Checks protected characters, world keys, and major character shrink.
 *
 * Replaces scripts/validate_patch.py.
 *
 * Usage:
 *   npx tsx skills/novel-writer/ts/validate-patch.ts \
 *     --project mynovel --patch patch.json
 */

import fsSync from "node:fs";
import path from "node:path";
import { parseArgs } from "node:util";

const PROJECTS_DIR = path.resolve(import.meta.dirname, "../projects");

type AnyObj = Record<string, any>;

function memPath(project: string, file: string): string {
  return path.join(PROJECTS_DIR, project, "memory", file);
}

function loadJson(filePath: string, fallback: unknown): any {
  if (!fsSync.existsSync(filePath)) return fallback;
  try {
    return JSON.parse(fsSync.readFileSync(filePath, "utf-8"));
  } catch {
    return fallback;
  }
}

function byName(items: AnyObj[]): Map<string, AnyObj> {
  const map = new Map<string, AnyObj>();
  for (const item of items) {
    const name = String(item.name ?? "").trim();
    if (name) map.set(name, item);
  }
  return map;
}

function validateCharacters(current: AnyObj, patch: AnyObj): void {
  const items: AnyObj[] = current.characters ?? [];
  const existing = byName(items);

  const toDelete = new Set<string>();
  for (const item of (patch.delete ?? []) as any[]) {
    const name = String(typeof item === "object" ? (item?.name ?? item) : item).trim();
    if (name) toDelete.add(name);
  }

  const protectedNames = new Set<string>();
  for (const [n, v] of existing) {
    if (v.protected === true) protectedNames.add(n);
  }

  const illegal = [...toDelete].filter((n) => protectedNames.has(n));
  if (illegal.length > 0) {
    throw new Error(`cannot delete protected characters: ${illegal.sort().join(", ")}`);
  }

  // Simulate patch to check major shrink
  for (const item of (patch.add ?? []) as AnyObj[]) {
    const name = String(item.name ?? "").trim();
    if (name) existing.set(name, { ...existing.get(name), ...item });
  }
  for (const item of (patch.update ?? []) as AnyObj[]) {
    const name = String(item.name ?? "").trim();
    if (name) existing.set(name, { ...existing.get(name), ...item });
  }
  for (const name of toDelete) {
    if (!protectedNames.has(name)) existing.delete(name);
  }

  const majorsBefore = items.filter((v) => v.role === "main" || v.protected);
  const majorsAfter = [...existing.values()].filter((v) => v.role === "main" || v.protected);
  if (
    majorsBefore.length > 0 &&
    majorsAfter.length < Math.max(1, Math.floor(majorsBefore.length * 0.7))
  ) {
    throw new Error("character shrink validation failed (major characters drop)");
  }
}

function validateWorld(current: AnyObj, patch: AnyObj): void {
  const protectedKeys = new Set<string>(current.protected_keys ?? []);
  const toDelete = (patch.delete ?? []) as string[];
  const illegal = toDelete.filter((key) => typeof key === "string" && protectedKeys.has(key));
  if (illegal.length > 0) {
    throw new Error(`cannot delete protected world keys: ${illegal.sort().join(", ")}`);
  }
}

function main() {
  const { values } = parseArgs({
    options: {
      project: { type: "string" },
      patch: { type: "string" },
    },
    strict: true,
  });

  if (!values.project || !values.patch) {
    console.error("--project and --patch are required");
    process.exit(1);
  }

  const patchFile = values.patch;
  if (!fsSync.existsSync(patchFile)) {
    console.error(`patch file not found: ${patchFile}`);
    process.exit(1);
  }

  const patchData: AnyObj = JSON.parse(fsSync.readFileSync(patchFile, "utf-8"));
  const characters = loadJson(memPath(values.project, "characters.json"), { characters: [] });
  const world = loadJson(memPath(values.project, "world_bible.json"), {
    world: {},
    protected_keys: [],
  });

  try {
    validateCharacters(characters, patchData.characters ?? {});
    validateWorld(world, patchData.world_bible ?? {});
    console.log(JSON.stringify({ status: "ok" }));
  } catch (err) {
    console.error(`Validation failed: ${err instanceof Error ? err.message : err}`);
    process.exit(1);
  }
}

main();
